src/java/org/apache/cassandra/index/sai/disk/v1/V1OnDiskFormat.java

```diff
@@ -0,0 +1,256 @@
+                                                                                 IndexComponent.TOKEN_VALUES,
+                                                                                 IndexComponent.PRIMARY_KEY_TRIE,
+                                                                                 IndexComponent.PRIMARY_KEY_BLOCKS,
+                                                                                 IndexComponent.PRIMARY_KEY_BLOCK_OFFSETS);
+    private static final Set<IndexComponent> LITERAL_COMPONENTS = EnumSet.of(IndexComponent.COLUMN_COMPLETION_MARKER,
+                                                                             IndexComponent.META,
+                                                                             IndexComponent.TERMS_DATA,
+                                                                             IndexComponent.POSTING_LISTS);
+
+    /**
+     * Global limit on heap consumed by all index segment building that occurs outside the context of Memtable flush.
+     *
+     * Note that to avoid flushing small index segments, a segment is only flushed when
+     * both the global size of all building segments has breached the limit and the size of the
+     * segment in question reaches (segment_write_buffer_space_mb / # currently building column indexes).
+     *
+     * ex. If there is only one column index building, it can buffer up to segment_write_buffer_space_mb.
+     *
+     * ex. If there is one column index building per table across 8 compactors, each index will be
+     *     eligible to flush once it reaches (segment_write_buffer_space_mb / 8) MBs.
+     */
+    public static final long SEGMENT_BUILD_MEMORY_LIMIT = 1024L * 1024L * DatabaseDescriptor.getSAISegmentWriteBufferSpace();
+
+    public static final NamedMemoryLimiter SEGMENT_BUILD_MEMORY_LIMITER =
+    new NamedMemoryLimiter(SEGMENT_BUILD_MEMORY_LIMIT, "SSTable-attached Index Segment Builder");
+
+    static
+    {
+        CassandraMetricsRegistry.MetricName bufferSpaceUsed = DefaultNameFactory.createMetricName(AbstractMetrics.TYPE, "SegmentBufferSpaceUsedBytes", null);
+        CassandraMetricsRegistry.Metrics.register(bufferSpaceUsed, (Gauge<Long>) SEGMENT_BUILD_MEMORY_LIMITER::currentBytesUsed);
+
+        CassandraMetricsRegistry.MetricName bufferSpaceLimit = DefaultNameFactory.createMetricName(AbstractMetrics.TYPE, "SegmentBufferSpaceLimitBytes", null);
+        CassandraMetricsRegistry.Metrics.register(bufferSpaceLimit, (Gauge<Long>) () -> SEGMENT_BUILD_MEMORY_LIMIT);
+
+        // Note: The active builder count starts at 1 to avoid dividing by zero.
+        CassandraMetricsRegistry.MetricName buildsInProgress = DefaultNameFactory.createMetricName(AbstractMetrics.TYPE, "ColumnIndexBuildsInProgress", null);
+        CassandraMetricsRegistry.Metrics.register(buildsInProgress, (Gauge<Long>) () -> SegmentBuilder.ACTIVE_BUILDER_COUNT.get() - 1);
```

Comment:
nit: I can't remember why we didn't just have `ACTIVE_BUILDER_COUNT` be what it is and floor to 1 for calculations like this. (I know we'd also have to do things like using `incrementAndGet()` in the `SegmentBuilder` constructor...)